libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/model/eh/EHorizonEdge.kt

```diff
@@ -0,0 +1,69 @@
+package com.mapbox.navigation.core.trip.model.eh
+
+/**
+ * Basic Edge class
+ *
+ * @param id identifier of the directed edge (it's unique)
+ * @param level the level of the Edge (0 being the mpp, 1 branches of the mpp,
+ * 2 branches of level 1 branches, etc)
+ * @param probability the probability for this edge in percentage to be taken by the driver.
+ * The probabilities of all outgoing edges on a single intersection sum up to 1.
+ * @param out the outgoing Edges
+ * @param parent the parent Edge
+ */
+class EHorizonEdge internal constructor(
+    val id: Long,
+    val level: Byte,
+    val probability: Double,
+    val out: List<EHorizonEdge>,
```

Comment:
`val out: List<EHorizonEdge>` is not presented in `equals` and  `hashCode`

Reply:
yep, with `out` in the hashCode I got StackOverflow when tried to add edges to the `Map`, so I removed it. 
maybe  I should add it again but without a recursive call.
`result = 31 * result + out.hashCode()` change to something like this -> `result = 31 * result + SOME_CONST`
or just add a comment and let it be removed.
wdyt?

Reply:
we cannot just leave a comment because hashCode/equals work incorrectly.

it's happening because of infinity invoking of hashCode fun. You can add check smt like
```
override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + level
        result = 31 * result + probability.hashCode()
        result = 31 * result + (parent?.hashCode() ?: 0)
        if(out.isNotEmpty()) {
              result = 31 * result + out.hashCode()
        }
        return result
    }
```

Reply:
`parent` calls hashCode of `out`s, each `out` call hashCode of `parent`. Infinity calls

Reply:
As a discussed option: write your own data structure(like `Tree` or use the existing one) and remove the `parent` property.

Reply:
discussed offline, as a quick fix we can use only `id` in hashCode and equals as it's unique.

Reply:
Noting that we run into the same issue and opted to use the `id` as well https://github.com/mapbox/mapbox-navigation-android/blob/834573c8f78781a239bef48cda852ae4720e39cf/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/model/eh/Edge.kt#L81 https://github.com/mapbox/mapbox-navigation-android/blob/834573c8f78781a239bef48cda852ae4720e39cf/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/model/eh/Edge.kt#L111

Reply:
@Guardiola31337 @RingerJK refactored `mpp()` method to use recursion. Don't need `parent` field to build `mpp`. Added tests. `hashCode` and `equals` contain all fields, no issues.